content/general/concepts/programming-optimization/programming-optimization.md

```diff
@@ -0,0 +1,243 @@
+
+Techniques used can be broken up among various scopes which can affect anything from a single statement to the entire program. Generally speaking, locally scoped techniques are easier to implement than global ones, though they may result in smaller gains. 
+
+These are important points to target for optimization in programming:
+
+- reducing code complexity
+- avoiding unnecessary for-loops
+- using data structures similar to like sorting, string concatenation
+- avoiding unnecessary functions
+- avoiding dots
+- initializing dictionary elements
+- import statement overhead
+- data aggregation
+- re-map functions at runtime
+- profiling code   
+
+## Levels of Optimization
+Every application, generally, is composed of 2 areas where programming optimizations can be applied within levels of development:
+
+***Client Side***, which is the side seen on the web browser by the user, includes: 
+- JavaScript that runs in the browser
+- load times for pages
+- images
+- videos
+- downloading of any resources required by the application, etc.
+
+A few of the techniques that can be applied optimize the programming on the client side could be: 
+- caching and content delivery networks
+- bundle and minification
+- image usage
+- removing duplicate JavaScript and CSS code
+- using minimalist styling frameworks, etc.
+
+***Server Side***, this second side relates to any computations done behind the scenes, tends to involve:
+- database queries and processing
+- application configuration and dependencies, etc.        
+
+Although not restricted to only these, typical strategies that help with this include:
+- using an appropriate server application instead of the default
+- enable caching(opcode, memory, http and application)
+- de-fragmentation of database tables
+- database settings
+- virtualization
+- dns query response time
+- and again, this list can go on depending how deep requirements are needed. 
+
+These programming optimizations can occur at different levels within the cycle of the application's creation. 
+
+Consideration tends to be given to efficiency throughout a project, though this varies significantly, but major optimization in the programming is often considered a refinement to be done later, if ever. 
+
+Typically the higher levels have greater impact and are harder to change later on in a project, requiring significant changes or even a complete rewrite. 
+
+On longer-running projects there are typically cycles of optimization, where improving one area reveals limitations in another, and these are typically curtailed when performance is acceptable and gains become too small or costly.
+
+### Design
+The architecture design of a system plays a major role, its optimization can overwhelmingly affect the system's performance by making the best use of the available resources and expected load (e.g. network latency can be optimized to minimize requests, ideally making a single request rather than multiple requests to the network).
```

Comment:
```suggestion

The architecture design of a system plays a major role, its optimization can overwhelmingly affect the system's performance by making the best use of the available resources and expected load (e.g. network latency can be optimized to minimize requests, ideally making a single request rather than multiple requests to the network).
```